### PR TITLE
feat(AWS HTTP API): Support `payload` version per function

### PR DESCRIPTION
**Pull request by nyamba**
please check implementation for https://github.com/serverless/serverless/issues/9402
Closes: #{9402}

### CODE REVIEW

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Hello @nyamba - thanks for the PR. It would be great to ensure that implemented feature works as intended by implementing corresponding tests, they should live in `httpApi.test.js` and use `runServerless`. Please let me know if you have any questions :bow:

**Comment by nyamba** [comment]
Hello @pgrzesik , okay I will add a test for it.

**Comment by nyamba** [comment]
@pgrzesik please check.

File: lib/plugins/aws/provider.js [hunk 1]
```diff
@@ -1230,6 +1230,13 @@ class AwsProvider {
             tracing: { $ref: '#/definitions/awsLambdaTracing' },
             versionFunction: { $ref: '#/definitions/awsLambdaVersioning' },
             vpc: { $ref: '#/definitions/awsLambdaVpcConfig' },
+            httpApi: {
+              type: 'object',
+              properties: {
+                payload: { type: 'string', enum: ['1.0', '2.0'] },
```
Comment (pgrzesik) [inline lib/plugins/aws/provider.js:1236]: let's stick to `payloadVersion` variable name to not introduce discrepancies between `provider.httpApi` and `function.httpApi` as it might be confusing to users that properties controlling the same thing are called differently

Comment (pgrzesik) [reply lib/plugins/aws/provider.js:1236]: Additionally, let's move the `payloadVersion` schema to a definition so it can be reused - it will help us ensure to not omit upading it in all places when we will need to change that definition in the future

File: test/integration/httpApi.test.js [hunk 2]
```diff
@@ -351,4 +351,36 @@ describe('HTTP API Integration Test', function () {
       expect(response.status).to.equal(404);
     });
   });
+
+  describe('Payload version per function', () => {
```
Comment (pgrzesik) [inline test/integration/httpApi.test.js:355]: I don't think we need an integration test for this one. Instead, let's configure `runServerless`-based tests here: https://github.com/serverless/serverless/blob/master/test/unit/lib/plugins/aws/package/compile/events/httpApi.test.js

Tests should check that the `httpApi` setting on function level is overriding setting from provider.

**Comment by nyamba** [comment]
thanks @pgrzesik for  detailed suggestions. 🙏
> let's stick to `payloadVersion` variable name to not introduce discrepancies between `provider.httpApi` and `function.httpApi` as it might be confusing to users that properties controlling the same thing are called differently

About the variable name,  `provider.httpApi` it has `payload` property, doesn't have `payloadVersion`. Please see 
lib/plugins/aws/provider.js:892
However, `provider.httpApi.authorizers` has `payloadVersion` property.
What should we call the variable name `payloadVersion` or `payload`? please correct me if I'm wrong :-)

**Comment by pgrzesik** [comment]
Sorry @nyamba - you're totally right, I've looked in the wrong place and the correct one is `payload` :+1: Thanks for being vigilant :bow:

**Comment by nyamba** [comment]
@pgrzesik please check. I've pushed the changes.

File: lib/plugins/aws/provider.js [hunk 3]
```diff
@@ -304,6 +304,10 @@ class AwsProvider {
             },
             additionalProperties: false,
           },
+          awsApiGatewayPayloadVersion: {
```
Comment (pgrzesik) [inline lib/plugins/aws/provider.js:307]: let's call it `awsHttpApiPayload` to keep the naming consistent with where it will be used

Comment (pgrzesik) [reply lib/plugins/aws/provider.js:307]: Also, let's use that definition inside `httpApi.payload` schema definition as it should be using the same one

File: test/unit/lib/plugins/aws/package/compile/events/httpApi.test.js [hunk 4]
```diff
@@ -125,6 +125,32 @@ describe('lib/plugins/aws/package/compile/events/httpApi.test.js', () => {
       const apiResource = cfResources[naming.getHttpApiLogicalId()];
       expect(apiResource.Properties.DisableExecuteApiEndpoint).to.equal(undefined);
     });
+
+    it('should support payload format version per function', async () => {
+      const { awsNaming, cfTemplate } = await runServerless({
```
Comment (pgrzesik) [inline test/unit/lib/plugins/aws/package/compile/events/httpApi.test.js:130]: We could avoid extra `runServerless` run if you could reuse the `runServerles` result that are setup in `before` step. Could you adjust configuration that is run in before so it will also support testing this case? Please see how tests above are configured

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thanks @nyamba :bow: Awesome job :100: I have only few minor comments, please let me know what do you think. Additionally, could you add a small note to: https://github.com/serverless/serverless/blob/master/docs/providers/aws/events/http-api.md#event--payload-format as well that it is now possible to set payload on function level?

**Comment by nyamba** [comment]
Hello @pgrzesik , I have pushed the improvements. Thanks for your time :)

File: docs/providers/aws/events/http-api.md [hunk 5]
```diff
@@ -392,6 +392,18 @@ provider:
...
 
+Also, the payload version could be configured in function level as follows:
+
+```yaml
+functions:
+  hello:
+    handler: index.handler
+    httpApi:
+      payload: '1.0'
+    method: GET
```
Comment (pgrzesik) [inline docs/providers/aws/events/http-api.md:403]: I believe the listing here is not correct it doesn't configure a valid event. 

It should rather look similar to:
```
functions:
  hello:
    handler: index.handler
    httpApi:
      payload: '1.0'
    events:
      - httpApi:
          path: /hello
          method: GET
```

File: docs/providers/aws/events/http-api.md [hunk 6]
```diff
@@ -392,6 +392,18 @@ provider:
     payload: '1.0'
 ```
 
+Also, the payload version could be configured in function level as follows:
```
Comment (pgrzesik) [inline docs/providers/aws/events/http-api.md:395]: Let's maybe reword it as

The payload version can also be specified at the function level with `httpApi.payload` property and it will take precedence over the payload version configured at the provider level. It can be configured as:

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Awesome job @nyamba :clap: I only have small suggestions related to docs and we're good to go :rocket:

**Comment by nyamba** [comment]
@pgrzesik Thanks for the correction :)

**Review by pgrzesik — APPROVED** [review]
Thanks @nyamba - it looks great :bow:
